### DISCUSSION TITLE
convert vs magick on windows server outputs different images.

### DISCUSSION THREAD

**Discussion by marchenault** [ImageMagick/ImageMagick#4175] [Help]
Notice the differences between these 2 commands when running from terminal

C:\Sites\Dependencies\ImageMagick-7.1.0-Q16-HDRI\magick koala.jpg -resize 765x574 -quality 92 -page 800x800 +repage -sharpen 0x.2 -background white -rotate 35 +repage -page +-182+-137 -background white -flatten -crop 160x180+40+135 output.jpg

C:\Sites\Dependencies\ImageMagick-7.1.0-Q16-HDRI\convert.exe koala.jpg -resize 765x574 -quality 92 -page 800x800 +repage -sharpen 0x.2 -background white -rotate 35 +repage -page +-182+-137 -background white -flatten -crop 160x180+40+135 output.jpg

it seems that the crop x & y are not where they should be when using magick.

Why are the same commands outputting different images? this is a real issue for us as we try to migrate to linux (AWS lambda) from windows.

do I need to write the commands differently? 
![Koala](https://user-images.githubusercontent.com/73128308/132738534-b78a7f72-608a-4c4d-91ff-abfd9e2fae81.jpg)

**Comment by marchenault** [comment]
Here are the results of each command.
![magick](https://user-images.githubusercontent.com/73128308/132741160-4113faa3-866c-461b-a81e-57c848d415f8.jpg)
![convert](https://user-images.githubusercontent.com/73128308/132741399-adee9983-9286-4daa-a7d6-c449b89efab6.jpg)

**Comment by snibgo** [comment]
Yes, various changes have occurred between v6 and v7. See http://www.imagemagick.org/script/porting.php

Relevant to you: 

1. "+repage" now sets the canvas dimensions to the image dimensions, in addition to setting the canvas offset to zero. 

2. "-flatten" now removes the page offset.

If you want v7 IM to process commands as if they were v6, use "magick convert" instead of just "magick".

The command you show seems somewhat random, hacked as if the author didn't know what they wanted or what the operations did. I recommend a re-write.

**Comment by marchenault** [comment]
Unfortunately using "magick convert" may work on windows but doesn't work for MagickCLI or GraphicsMagick so this is not a viable solution for porting the app to AWS lambda

C:\Sites\Dependencies\ImageMagick-7.1.0-Q16-HDRI\convert.exe koala.jpg -resize 765x574 -quality 92 -page 800x800 +repage -sharpen 0x.2 -background white -rotate 35 +repage -page +-182+-137 -background white -flatten -crop 160x180+40+135 output.jpg 

The output above is exactly what we want. The numeric values apart from 800x800 are all created by our UI interface. This interface allows users to scale, rotate and crop.

seems to me I need to change the syntax to magick somehow.

**Comment by snibgo** [comment]
GraphicsMagick is a different system to ImageMagick.

I know nothing about AWS lambda.

The following (Windows syntax) gives the same result in IM v7.1.0-4 using "magick" or "magick convert", or v6.9.9-50 using "convert", and the same as your result from v6 "convert".
```
magick ^
  koala.jpg ^
  -resize 765x574 ^
  -sharpen 0x.2 ^
  -background white -rotate 35 +repage ^
  -crop 160x180+404+409 +repage ^
  -quality 92 ^
  out.jpg
```

**Reply by marchenault** [reply]
I appreciate your help but some commands are missing. I need to have -page and rotate. I attached is our interface. notice how the image is not only rotated but is pull to the left, outside the bounding box, and then the crop box is off center. The image and crop box can be moved independently of each other. That is why I needed "-rotate 35 +repage -page +-182+-137"

![Screen Shot 2021-09-09 at 3 46 56 PM](https://user-images.githubusercontent.com/73128308/132772642-3d9d87fa-1a54-4972-9353-df1954088e06.png)

I need the same output of this: 

C:\Sites\Dependencies\ImageMagick-7.1.0-Q16-HDRI\convert.exe koala.jpg -resize 765x574 -quality 92 -page 800x800 +repage -sharpen 0x.2 -background white -rotate 35 +repage -page +-182+-137 -background white -flatten -crop 160x180+40+135 output.jpg

using magick instead of convert:

**Comment by marchenault** [comment]
I think I figured it out. if the user rotates the image the actual image becomes bigger to accommodate. so all I need to do is add or subtract from the crops xy.

This seems to work.
-resize 400x300 -sharpen 0x.2 -background white -rotate 36 +repage -crop 160x180+64+174 +repage -quality 92 

Thank you

**Comment by marchenault** [comment]
Unfortunately that didn't work. I need to be able to retain -rotate 35 +repage -page +-182+-137 -background white -flatten

**Comment by marchenault** [comment]
so back to my original question: how can this be rewritten so it will run with "magick"

**Comment by marchenault** [comment]
the final image could be resized, rotated and offset and then cropped. This works perfect in convert.

 -resize 765x574 -sharpen 0x.2 -page 800x800 +repage -background white -rotate 94 +repage -page +-305+-386 -flatten -crop 160x180-305-212 +repage -quality 92

In "magick" I get a white image.

**Comment by marchenault** [comment]
it seems to me that in "magick" -page doesn't do anything. I need for the image to be off the canvas like so:
![convert](https://user-images.githubusercontent.com/73128308/132778611-315a0953-8942-48d3-ae75-290c298a53bf.jpg)

any ideas?

**Comment by snibgo** [comment]
My command above gives:
![k3](https://user-images.githubusercontent.com/33812368/132782262-15ab0372-460d-4fa0-9cbb-89b0d22a71b1.jpg)

Isn't that what you want? It is identical to the v6 result you showed.

A more direct (and faster) command would use "-distort SRT" with a viewport.

Yes, you have more numbers in your command. I don't know what they represent. Perhaps they are numbers that come out of your UI. Perhaps you need to do some arithmetic to get the numbers for the IM command. Perhaps you can do that arithmetic in "%[fx:...]" expressions.

**Comment by marchenault** [comment]
my ui has a number of values:

1. resize: allows the user to zoom in and out of an image
2. rotation: 0-360 degreess
3. offset of the image in relation to the view port in x & y.
4. offset of the crop frame in x & y in relation to the view port.

I think the problem is that if the image is rotated and or zoomed (ie resized) I no longer know where to place the crop x and y.

Thank you for your help. it is very much appreciated.

**Comment by marchenault** [comment]
I have changed the logic so the crop xy coordinates now work with an image that was created without the use of -page and -flatten and it works as intended.
